primitives/state-machine/src/overlayed_changes/changeset.rs

```diff
@@ -0,0 +1,731 @@
+use smallvec::SmallVec;
+
+const PROOF_OVERLAY_NON_EMPTY: &str = "\
+	An OverlayValue is always created with at least one transaction and dropped as soon
+	as the last transaction is removed; qed";
+
+type DirtyKeys = SmallVec<[HashSet<StorageKey>; 5]>;
+type Transactions = SmallVec<[InnerValue; 5]>;
+
+/// Error returned when trying to commit or rollback while no transaction is open.
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct NoOpenTransaction;
+
+/// Describes in which mode the node is currently executing.
+#[derive(Debug, Clone, Copy)]
+pub enum ExecutionMode {
+	/// Exeuting in client mode: Removal of all transactions possible.
+	Client,
+	/// Executing in runtime mode: Transactions started by the client are protected.
+	Runtime,
+}
+
+/// Describes which kind of transaction close should be performed.
+#[derive(Debug, Clone, Copy)]
+enum CloseType {
+	/// Commit the transaction.
+	Commit,
+	/// Rollback the transaction.
+	Rollback,
+}
+
+#[derive(Debug, Default, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+struct InnerValue {
+	/// Current value. None if value has been deleted.
+	value: Option<StorageValue>,
+	/// The set of extrinsic indices where the values has been changed.
+	/// Is filled only if runtime has announced changes trie support.
+	extrinsics: BTreeSet<u32>,
+}
+
+/// An overlay that contains all versions of a value for a specific key.
+#[derive(Debug, Default, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct OverlayedValue {
+	/// The individual versions of that value.
+	/// One entry per transactions during that the value was actually written.
+	transactions: Transactions,
+}
+
+/// Holds a set of changes with the ability modify them using nested transactions.
+#[derive(Debug, Default, Clone)]
+pub struct OverlayedChangeSet {
+	/// Stores the changes that this overlay constitutes.
+	changes: BTreeMap<StorageKey, OverlayedValue>,
+	/// Stores which keys are dirty per transaction. Needed in order to determine which
+	/// values to merge into the parent transaction on commit. The length of this vector
+	/// therefore determines how many nested transactions are currently open (depth).
+	dirty_keys: DirtyKeys,
+	/// The number of how many transactions beginning from the first transactions are started
+	/// by the client. Those transactions are protected against close (commit, rollback)
+	/// when in runtime mode.
+	num_client_transactions: usize,
+	/// Determines whether the node is using the overlay from the client or the runtime.
+	execution_mode: ExecutionMode,
+}
+
+impl Default for ExecutionMode {
+	fn default() -> Self {
+		Self::Client
+	}
+}
+
+impl OverlayedValue {
+	/// The value as seen by the current transaction.
+	pub fn value(&self) -> Option<&StorageValue> {
+		self.transactions.last().expect(PROOF_OVERLAY_NON_EMPTY).value.as_ref()
+	}
+
+	/// Unique list of extrinsic indices which modified the value.
+	pub fn extrinsics(&self) -> impl Iterator<Item=&u32> {
```

Comment:
Is this function used exclusively in tests? If so do you think we could hide it behind `cfg(test)`? If so, I see an opportunity to hide `itertools` under dev-dependencies as well.

Reply:
No. It it used outside of tests, too. It is used in `prepare_extrinsics_input_inner`.